source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -127,19 +127,48 @@ void SimpleHttpCache::updateHeaders(const LookupContext& lookup_context,
     return;
   }
 
-  // https://www.rfc-editor.org/rfc/pdfrfc/rfc7234.txt.pdf
-  // 4.3.4 Freshening Stored Responses upon Validation
-  // use other header fields provided in the 304 (Not Modified)
-  // response to replace all instances of the corresponding header
-  // fields in the stored response.
-  //
+  // Skip headers that we don't want to update because of specific reasons
+  static const auto* headersNotToUpdate = new absl::flat_hash_set<Http::LowerCaseString>({
+      // Content range should not be changed upon validation
+      Http::Headers::get().ContentRange,
+
+      // The age is calculated and set by the general cache_filter code logic.
+      Http::CustomHeaders::get().Age,
+
+      // Headers that describe the body content should never be updated.
+      Http::Headers::get().ContentLength,
+      Http::Headers::get().ContentType,
+
+      // It does not make sense for this level of the code to be updating the ETag, when presumably
+      // the
+      // cached_response_headers reflect this specific ETag.
+      Http::CustomHeaders::get().Etag,
+
+      // We don't update the cached response on a Vary; we just delete it
+      // entirely. So don't bother copying over the Vary header.
+      Http::CustomHeaders::get().Vary,
+  });
+
   // Assumptions:
   // 1. The internet is fast, i.e. we get the result as soon as the server sends it.
-  // Race conditions would not be possible because we are always processing up-to-date data.
+  //    Race conditions would not be possible because we are always processing up-to-date data.
   // 2. No key collision for etag. Therefore, if etag matches it's the same resource.
   // 3. Backend is correct. etag is being used as a unique identifier to the resource
-  // TODO(tangsaidi) merge the header map instead of replacing it according to rfc7234
-  entry.response_headers_ = Http::createHeaderMap<Http::ResponseHeaderMapImpl>(response_headers);
+
+  // use other header fields provided in the new response to replace all instances
+  // of the corresponding header fields in the stored response
+  // Note: Warning 1xx, 2xx are not handled as they are not commonly used despite
```

Comment:
perhaps we should say "not handled yet" as we wouldn't be against someone adding support for them, but we just don't care to spend time on it right now.

Reply:
It looks like according to the latest [IETF doc](https://www.ietf.org/archive/id/draft-ietf-httpbis-cache-18.html) recognizes the warning codes are obsolete and removed it from the specs. I'll probably remove this note altogether.